File: Backend/src/main/java/com/sencarmarket/module/admin/mapper/AdminMapper.java
package com.sencarmarket.module.admin.mapper;

import com.sencarmarket.module.paiement.dto.TransactionResponse;
import com.sencarmarket.module.paiement.entity.TransactionPortefeuille;
import com.sencarmarket.module.utilisateur.dto.UtilisateurResponse;
import com.sencarmarket.module.utilisateur.entity.Utilisateur;
import com.sencarmarket.module.vehicule.dto.VehiculeResponse;
import com.sencarmarket.module.vehicule.entity.Vehicule;
import org.springframework.stereotype.Component;

import java.util.UUID;

/**
 * Mapper pour les conversions Admin - Respecte le principe DRY
 * Single Responsibility: только маппинг
 */
@Component
public class AdminMapper {

    // ==================== UTILISATEUR ====================

    public UtilisateurResponse toUtilisateurResponse(Utilisateur utilisateur) {
        if (utilisateur == null) {
            return null;
        }

        String typeRole = utilisateur.getTypeUtilisateur() != null 
                ? utilisateur.getTypeUtilisateur().getNom() 
                : null;
        String statutVerif = utilisateur.getStatutVerification();

        return UtilisateurResponse.builder()
                .id(utilisateur.getId())
                .email(utilisateur.getEmail())
                .nom(utilisateur.getNom())
                .prenom(utilisateur.getPrenom())
                .telephone(utilisateur.getTelephone())
                .typeUtilisateur(typeRole)
                .statutVerification(statutVerif)
                .emailVerifie(utilisateur.getEmailVerifie())
                .telephoneVerifie(utilisateur.getTelephoneVerifie())
                .doubleAuthActive(utilisateur.getDoubleAuthActive())
                .createdAt(utilisateur.getCreatedAt())
                .build();
    }

    // ==================== VEHICULE ====================

    public VehiculeResponse toVehiculeResponse(Vehicule vehicule) {
        if (vehicule == null) {
            return null;
        }

        String marqueName = vehicule.getMarque() != null 
                ? vehicule.getMarque().getNom() 
                : null;
        String modeleName = vehicule.getModele() != null 
                ? vehicule.getModele().getNom() 
                : null;
        String statutName = vehicule.getStatut() != null 
                ? vehicule.getStatut().name() 
                : null;

        return VehiculeResponse.builder()
                .id(vehicule.getId())
                .marque(marqueName)
                .modele(modeleName)
                .anneeFabrication(vehicule.getAnneeFabrication())
                .kilometrage(vehicule.getKilometrage())
                .couleur(vehicule.getCouleur())
                .prixVente(vehicule.getPrixVente())
                .description(vehicule.getDescription())
                .numeroVin(vehicule.getNumeroVin())
                .immatriculation(vehicule.getImmatriculation())
                .statut(statutName)
                .prixNegociable(vehicule.getPrixNegociable())
                .certifie(vehicule.getCertifie())
                .estBoost(vehicule.getEstBoost())
                .boostDebut(vehicule.getBoostDebut())
                .boostFin(vehicule.getBoostFin())
                .vues(vehicule.getVues())
                .createdAt(vehicule.getCreatedAt())
                .build();
    }

    // ==================== TRANSACTION ====================

    public TransactionResponse toTransactionResponse(TransactionPortefeuille transaction) {
        if (transaction == null) {
            return null;
        }

        UUID portefeuilleId = transaction.getPortefeuille() != null 
                ? transaction.getPortefeuille().getId() 
                : null;

        return TransactionResponse.builder()
                .id(transaction.getId())
                .portefeuilleId(portefeuilleId)
                .montant(transaction.getMontant())
                .typeTransaction(transaction.getTypeTransaction())
                .statut(transaction.getStatut())
                .description(transaction.getDescription())
                .referenceExterne(transaction.getReferenceExterne())
                .dateTransaction(transaction.getDateTransaction())
                .createdAt(transaction.getCreatedAt())
                .build();
    }
}

File: Backend/src/modules/admin/services/admin.mapper.ts
import { Injectable } from '@nestjs/common';

import { toNumberOrZero } from '../../../common/utils/number.util';
import { UtilisateurResponseDto } from '../../auth/dto/utilisateur-response.dto';
import { TransactionResponseDto } from '../../paiement/dto/transaction-response.dto';
import { VehiculeResponseDto } from '../../vehicule/dto/vehicule-response.dto';
import { AdminUserRecord, TransactionRecord, VehiculeRecord } from '../admin.models';

@Injectable()
export class AdminMapper {
  toUtilisateurResponse(user: AdminUserRecord): UtilisateurResponseDto {
    return {
      id: user.id,
      email: user.email,
      telephone: user.telephone,
      prenom: user.prenom,
      nom: user.nom,
      photoProfilUrl: user.photoProfilUrl,
      emailVerifie: user.emailVerifie,
      telephoneVerifie: user.telephoneVerifie,
      doubleAuthActive: user.doubleAuthActive,
      typeUtilisateur: user.typeUtilisateur?.nom ?? null,
      statutVerification: user.statutVerification,
      createdAt: user.createdAt,
    };
  }

  toVehiculeResponse(vehicule: VehiculeRecord): VehiculeResponseDto {
    return {
      id: vehicule.id,
      marque: vehicule.marque?.nom ?? null,
      marqueId: vehicule.marqueId,
      modele: vehicule.modele?.nom ?? null,
      modeleId: vehicule.modeleId,
      anneeFabrication: vehicule.anneeFabrication,
      kilometrage: vehicule.kilometrage,
      carburant: vehicule.carburant?.nom ?? null,
      carburantId: vehicule.carburantId,
      boiteVitesse: vehicule.boiteVitesse?.nom ?? null,
      boiteVitesseId: vehicule.boiteVitesseId,
      couleur: vehicule.couleur,
      prixVente: vehicule.prixVente != null ? String(vehicule.prixVente) : null,
      description: vehicule.description,
      numeroVin: vehicule.numeroVin,
      immatriculation: vehicule.immatriculation,
      statut: vehicule.statut,
      prixNegociable: vehicule.prixNegociable,
      certifie: vehicule.certifie,
      photosUrls: vehicule.photos.map((photo) => photo.url),
      estBoost: vehicule.estBoost,
      boostDebut: vehicule.boostDebut,
      boostFin: vehicule.boostFin,
      vues: vehicule.vues,
      nombreFavoris: vehicule.nombreFavoris,
      estFavori: false,
      proprietaireNom: vehicule.proprietaire?.nom ?? null,
      proprietaireId: vehicule.proprietaireId,
      createdAt: vehicule.createdAt,
    };
  }

  toTransactionResponse(transaction: TransactionRecord): TransactionResponseDto {
    return {
      id: transaction.id,
      portefeuilleId: transaction.portefeuilleId,
      montant: String(toNumberOrZero(transaction.montant)),
      typeTransaction:
        (transaction.typeTransaction as TransactionResponseDto['typeTransaction']) ?? 'CREDIT',
      statut: (transaction.statut as TransactionResponseDto['statut']) ?? 'EN_ATTENTE',
      description: transaction.description,
      referenceExterne: transaction.referenceExterne,
      dateTransaction: transaction.dateTransaction,
      createdAt: transaction.createdAt,
    };
  }
}
